text/0072-metric-observer.md

```diff
@@ -0,0 +1,202 @@
+will be able to use these metrics directly.  Non-monotonic Counters
+may be expressed as their current value, but they cannot meaningfully
+be aggregated in this way.
+
+The preferred way to `Observe()` Counter-like data from an Observer
+instrument callback is to report deltas in the callback and configure
+a Sum aggregation in the exporter.  Data reported in this way will
+support rate calculations just as if they were true Counters.
+
+### Differences between Gauge and Observer
+
+One significant difference between gauges that are explicitly `Set()`,
+as compared with observer callbacks, is that `Set()` happens inside a
+context (i.e., its distributed context), whereas the observer callback
+does not execute with any distributed context.  
+
+Whereas Gauge values do have context at the moment `Set()` is called,
+Observer callbacks do not.  Observer instruments are appropriate for
+reporting values that are not request specific.
+
+## Details
+
+Observer instruments are semantically equivalent to Gauge instruments
+but use different calling conventions.  Use the language-specific
+constructor for an Observer instrument (e.g.,
+`metric.NewFloat64Observer()`).  Observer instruments support the
+`Monotonic` and `NonMonotonic` options, the same as Gauge instruments.
+
+Callbacks should avoid blocking.  The implementation may be required
+to cancel computation if the callback blocks for too long.
+
+Callbacks must not be called synchronously with application code via
+any OpenTelemetry API.  This prevents the application from potentially
+deadlocking itself by being called synchronously from its own thread.
+Implementations that cannot provide this guarantee should prefer not
+to implement Observer instrsuments.
+
+Callbacks may be called synchronously in the SDK on behalf of an
+exporter, provided it does not contradict the requirement above.
+
+Callbacks should avoid calling OpenTelemetry APIs other than the
+interface provided to `Observe()`.  This prevents the SDK from
+potentially deadlocking itself by being called synchronously from its
+own thread.  We recognize this may be impossible or expensive to
+enforce.  SDKs should document how they respond to such attempts at
+re-entry.
+
+### Observer calling conventions
+
+Observer callbacks accept a Result interface, which supports both
+bound and direct calls as follows.
+
+For a "direct" observation with a specific LabelSet, call the Result
+directly using `Result.Observe(value, LabelSet)` to report an
+observation.
+
+For a "bound" observation with using a bound observer instrument,
+first `Bind()` the instrument with a LabelSet, then call the bound
+instrument passing the result: `Result.ObserveBound(value,
+BoundInstrument)`.  Callbacks MUST use bound instruments corresponding the
+Observer instrument for which they are registered.  It is an error if
+`Result.ObserveBound(value, BoundInstrument)` is called for a Result
+corresponding to a different Observer instrument.
```

Comment:
As I understand it, "bound" metric instruments are an optimization to avoid repeatedly updating a metric for a particular `labelSet`. This saves us the cost of constructing the `labelSet` and looking up the aggregation to update, since the bound instrument has a reference to the aggregation.

Does this optimization still make sense in the context of callbacks? For non-observer metrics, we wouldn't expect the user to create a bound instrument for a single call.

Reply:
You ask a good question (also below), because if we had one callback per bound Observer/LabelSet, then there's no need to actually expose a Bound Instrument type. There would be a `Observer.Bind(LabelSet, callback)` method to bind a callback to a LabelSet instead, and a callback would be called once per bound label set and it could return a value instead of calling `ObserverResult` callback. This is "case (b)" in the updated language. Potentially this approach can evaluate callbacks in parallel, although that could also cause problems.

This confusion is what led to leaving Observers out of the v0.2 spec. I came away thinking that one callback per bound instrument makes sense, but if we intend a direct calling convention for Observer instruments, then having one callback per instrument gives a smaller surface area (and simpler SDK).  I'm on the fence here.

Reply:
I would suggest one callback per instrument, the main reason is that you can simply implement a callback that updates multiple labelsets, but if you have a callback per labelset and you need to access a system information you may end up doing a lot of unnecessary system calls. See the java example for memory.

I think @c24t is asking to completely remove bounds from the observer metrics because the benefit of bounding is unclear if the callback is called rarely (once per export interval) which I agree with and we can always add the bounds later to the Observers if we prove the performance wrong.

Reply:
It's a good point. The point of the bound instrument is to avoid a map lookup for each operation, but that map lookup isn't needed for observer instruments--they can bypass the map of current records. They observer instrument's observations are output directly, and since the Batcher (in the current arch diagram--I want to rename this) is required to support duplicates I expect no benefit from bound observer instruments.

Reply:
I have removed "bound" observer instruments based on this reasoning.

Reply:
I think it would be a move in the wrong direction. IMO, the callback use case is far more obscure than bound instruments (which are used all over the place in Jaeger code, for example). Bound instruments allow the code to provide measurements on its own pace, usually based on the business traffic. If callback is the only way to provide those, then user code needs to be accumulating the measurements between callbacks, a serious complication.

Reply:
The (now removed) case of a bound observer instrument was not meant as a stand-in for the user to accumulate measurements, but for a group of fixed label sets to be repeatedly recorded on the collection interval.

I agree completely with your statement -- Observer instruments should not be used when it involves the user accumulating measurements.  There should not be any benefit or performance advantage for the user because the SDK is designed and heavily optimized to do just this -- accumulate measurements. Bound instruments should be used when recording measurements. 

Observer instruments are meant for "computed values", which are passively collected measurements, not accumulated by the user but computed on the collection interval. The observation (which I credit to @c24t above) is that binding a label set with an observer instrument does not offer any performance benefit, because the emitted values will be directly output to the export pipeline.  They are not buffered or accumulated in the SDK, therefore binding them does no good.

I'm leaving this issue resolved, since we have removed bound observer instruments from this proposal.